keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1291 @@
+  Pod
+- solve a Job completion issue when sidecars should run continuously
+- allow mixing initContainers and sidecars for a choreographed startup sequence
+- allow easy injection of sidecar containers in any Pod
+
+### Non-Goals
+
+This proposal doesn't aim to:
+- support arbitrary dependencies graphs between containers
+- act as a security control to enforce that pod containers only run while the
+  sidecar is healthy. Restart of sidecar containers is a best effort
+- allow to enforce security boundaries to sidecar containers different that
+  other containers. For example, allow Istio to run privileged to configure ip
+  tables and disable other containers from doing so
+- (alpha) support containers termination ordering
+
+## Proposal
+
+The proposal is to introduce a `restartPolicy` field to Init containers and use
+it to indicate that an initContainer is a sidecar container. The only supported
+value proposed in this KEP is `Always`. Kubelet will start these containers in
+the order with other Init containers, but instead of waiting for its completion,
+it will wait for the container startup completion (startup probe succeeded or if
+no startup probe defined, postStart completed).
+
+These containers will not block Pod completion - if all regular containers
+complete, sidecar containers will be terminated.
+
+During the sidecar startup stage it will behave similarly to Init containers. If
+Pod `restartPolicy` is `Never`, sidecar container failed during startup will NOT
+be restarted. If Pod `restartPolicy` is `Always` or `OnFailure`, it will be
+restarted.
+
+Once sidecar container is started (postStart completed and startup probe
+succeeded), this containers will be restarted even when the Pod `restartPolicy`
+is `Never` or `OnFailure`. Furthermore, sidecar containers will be restarted
+even during Pod termination.
+
+In order to minimize OOM kills of sidecar containers, the OOM adjustment for
+these containers will match or exceed the OOM score adjustment of regular
+containers in the Pod. This intent to solve the issue
+https://github.com/kubernetes/kubernetes/issues/111356 
+
+As part of this KEP we also will be enabling for sidecar containers:
+- PreStart and PostStop lifecycle handlers for sidecar containers
+- All probes (startup, readiness, liveness)
+- Readiness probes of sidecars will be used to determine the whole Pod
+  readiness, but only after the main containers initialization started.
+  Readiness probes will not start running before the Pod initialization stage
+  completed 
+
+```yaml
+kind: Pod
+spec:
+  initContainers:
+  - name: vault-agent
+    image: hashicorp/vault:1.12.1
+    restartPolicy: OnFailure
+  - name: istio-proxy
+    image: istio/proxyv2:1.16.0
+    args: ["proxy", "sidecar"]
+    restartPolicy: Always
+  containers:
+  ...
+```
+
+### Future use of `restartPolicy` field
+
+In the Alpha version, no other values beyond Always will be acceptable. Going
```

Comment:
I agree. This section is written with some assumptions after thinking about it for a while. I think the main message here is that this field may make sense on other containers going forward.

How detailed do you see this message outlined? I really don't want to get into arguments on how this will work on other containers beyond "it will make sense for other containers/values" in future.

Reply:
I'd move ALL the future stuff to a later section and then you can include details and not have it be so distracting :)

Reply:
Moved all future stuff to later section